### PR TITLE
Minimal enclosing sphere fix

### PR DESCRIPTION
**Pull request by nigeljw-unity**
Chicken bit for the minimal enclosing sphere frustum cascade culling stabilization fix, and also user control over the  number of iterations for the iterative method for finding the minimal enclosing sphere from the 8 frustum cascade points. This fix makes it so that the culling spheres appropriately fully enclose the cascade. Since this is an iterative method, it is conservative. This fixes URP so there are not false positives when culling shadows.

image

https://jira.unity3d.com/browse/URP-782

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR

Fixes bug where shadows disappear in the frustum corners from false positives in the shadow culling.
https://fogbugz.unity3d.com/f/cases/1153151

culling_bug

Also fixes inherent bug with shadow fade where the shadows will be culled with a hard line across the fade which is due to the same problem.

CullingSphere

---
### Testing status
Tested with URP default template and custom cascade positioning.

culling_fix

---
### Comments to reviewers

This PR is dependent on the following change to trunk:
https://github.cds.internal.unity3d.com/unity/unity/pull/2418

Design doc:
https://docs.google.com/document/d/10YxFKdA8YFS7QS8Owg2dvnYaM-t9tsLXWQLtiEXZQvQ

### CODE REVIEW

**Comment by nigeljw-unity** [comment]
I still need to pull master into universal/debug-cull-features and set as the base branch to simplify the diff for review.

File: com.unity.render-pipelines.universal/Runtime/RendererFeatures/DebugCullingFeature.cs [hunk 1]
```diff
@@ -128,6 +128,23 @@ public static void DrawFrustumSplits(Matrix4x4 projMatrix, float splitMaxPct, Ve
             Vector4[] q = new Vector4[4];
             for (int j = splitStart; j < splitCount; j++)
             {
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/RendererFeatures/DebugCullingFeature.cs:130]: You can add the additions especially if you do it in a parametrized way.
Here I would've just passed an array of colors in and picked array[i % arrray.length] color. This way you could just keep using 1 color too.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 2]
```diff
@@ -178,6 +178,8 @@ public partial class UniversalRenderPipelineAsset : RenderPipelineAsset, ISerial
         [SerializeField] float m_ShadowDepthBias = 1.0f;
         [SerializeField] float m_ShadowNormalBias = 1.0f;
         [SerializeField] bool m_SoftShadowsSupported = false;
+        [SerializeField] bool m_TightEnclosingSphere = false;
+        [SerializeField] int m_NumIterationsEnclosingSphere = 64;
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:182]: How is this determined. Are there some sort of error bounds or empirical testing?

**Review by eh-unity — CHANGES_REQUESTED** [review]
I think debugCullingFeature shouldn't end up in master. It should be removed.
Otherwise seems good.

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Looking good! I've not found any issues besides what we discussed, so approving.

**Review by aleksandrasdzikia — APPROVED** [review]
The issue is no longer reproducible. Tested URP Examples scenes, looks good. The final result looks similar to desktop version, no artifacts found.

**Comment by nigeljw-unity** [comment]
@Unity-Technologies/gfx-qa-urp could I get a performance analysis of this change? I am curious if this potentially improves the rendering time of shadows due to the cascade culling sphere overlap removal.

**Review by erikabar — APPROVED** [review]
URP QA got re-added again, approving

**Comment by nigeljw-unity** [comment]
Adding the test results doc:
https://docs.google.com/document/d/1Sa5mLkJ8_NyBT9dh8hMlJnaVri1sQMtxKsbpxlv6no4

**Comment by nigeljw-unity** [comment]
@simon-engelbrecht-soerensen @erikabar would either of you be able to do a quick sanity test of this latest PR? None of the logic changed so no need to test in depth, but there was a big merge conflict that I had to resolve.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 2 above)
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:202]: Just some manual empirical testing, and this gets the accuracy down to the 0.001.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 3]
```diff
@@ -260,6 +262,9 @@ public static UniversalRenderPipelineAsset Create(ScriptableRendererData rendere
             // Initialize default Renderer
             instance.m_EditorResourcesAsset = instance.editorResources;
 
+            // Only enable for new URP assets by default
+            instance.m_ConservativeEnclosingSphere = true;
```
Comment (nigeljw-unity) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:266]: @Verasl @phi-lira @eh-unity This is how I am enabling it only for new projects. Let me know if this is the right approach.

**Review by eh-unity — APPROVED** [review]
lgtm.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAsset/UniversalRenderPipelineAssetUI.Skin.cs [hunk 4]
```diff
@@ -74,6 +74,7 @@ public static class Styles
             public static GUIContent shadowDepthBias = EditorGUIUtility.TrTextContent("Depth Bias", "Controls the distance at which the shadows will be pushed away from the light. Useful for avoiding false self-shadowing artifacts.");
             public static GUIContent shadowNormalBias = EditorGUIUtility.TrTextContent("Normal Bias", "Controls distance at which the shadow casting surfaces will be shrunk along the surface normal. Useful for avoiding false self-shadowing artifacts.");
             public static GUIContent supportsSoftShadows = EditorGUIUtility.TrTextContent("Soft Shadows", "If enabled pipeline will perform shadow filtering. Otherwise all lights that cast shadows will fallback to perform a single shadow sample.");
+            public static GUIContent conservativeEnclosingSphere = EditorGUIUtility.TrTextContent("Conservative Enclosing Sphere", "If enabled, a conservative enclosing sphere is used. If disabled, the sphere will be smaller than the cascade size, and shadows will be culled in the corners of the cascades.");
```
Comment (oleks-k) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAsset/UniversalRenderPipelineAssetUI.Skin.cs:77]: My suggestion:
EditorGUIUtility.TrTextContent("Conservative Enclosing Sphere", "Enable this option to improve shadow frustum culling. Enabling this option prevents Unity from excessively culling shadows in the corners of the shadow cascades. Disable this option only for compatibility purposes of projects created in previous Unity versions.");

**Review by oleks-k — CHANGES_REQUESTED** [review]

**Review by oleks-k — APPROVED** [review]
Looks good!

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 5]
```diff
@@ -913,6 +918,18 @@ public bool useAdaptivePerformance
             set { m_UseAdaptivePerformance = value; }
         }
 
+        public bool conservativeEnclosingSphere
+        {
+            get { return m_ConservativeEnclosingSphere; }
+            set { m_ConservativeEnclosingSphere = value; }
+        }
+
+        public int numItertionsEnclosingSphere
```
Comment (sandy-carter-unity) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:927]: A bit late to the party but there's a small spelling mistake here:

numItertionsEnclosingSphere -> numIterationsEnclosingSphere

**Pull request merged**